fairscale/optim/adascale.py

```diff
@@ -199,7 +199,11 @@ def _final_callback(self) -> None:
         # gradients have been synchronized between each worker.
         self._final_callback_queued = False
         assert isinstance(self._local_grad_sqr, torch.Tensor)
-        torch.distributed.all_reduce(self._local_grad_sqr / self._world_size)
+
+        # self._local_grad_sqr is FP32, sum then div shouldn't overflow.
+        torch.distributed.all_reduce(self._local_grad_sqr)  # SUM
```

Comment:
I'm actually curious, why does the result differ from before ?

Reply:
yeah, it is a subtle bug. The code before did the all_reduce but then the results were in a temp tensor, which is the result of self._local_grad_sqr / world_size. Therefore, local_grad_sqr didn't get updated as intended.

Reply:
ahh, makes sense, I was wondering whether that was correct when reading that. And how about
`self._local_grad_sqr.div_(self._world_size)
torch.distributed.all_reduce(self._local_grad_sqr)
`
?

Reply:
Right, I considered that briefly. I think this is a small tensor, not a float value per param. So it is easy to just keep it in FP32 and FP32's range is large enough to not overflow. If we do FP16 for it in the future, I think it might be worth doing the alternative.